tutorials/iot/graphical-snaps/graphical-snaps.md

```diff
@@ -0,0 +1,608 @@
+tags: snap, digital-signage, kiosk, device
+difficulty: 3
+published: 2018-05-17
+author: Gerry Boland <gerry.boland@canonical.com>
+
+---
+
+# Graphical Snaps for Ubuntu Core
+
+## Overview
+duration: 1:00
+
+This is a guide on how to create graphical snaps for Ubuntu Core with a single GUI application running fullscreen on the display. This addresses situations like:
+* Digital signage
+* Web kiosk
+* Industrial machine User Interface
+
+negative
+: This approach can be expanded to create more complex user interfaces with multiple GUI applications. However, this will require more advanced window management and is considered outside the scope of this guide.
+
+positive
+: The combination of Snap, the "mir-kiosk" Wayland server and Ubuntu Core ensures reliability and security of any graphical embedded device application.
+
+### What you'll learn
+
+How to create graphical snaps for Ubuntu Core using a toolkit that supports Wayland.
+
+### What you'll need
+
+You can follow the steps in this tutorial on any current release of Ubuntu. You'll also need a device (or a VM) with Ubuntu core installed.
+
+## Using Wayland
+duration: 3:00
+
+Graphics on Ubuntu Core uses Wayland as the primary protocol?. Mir is a graphical display server that supports Wayland clients. Snapd supports Wayland as an interface, so confinement can be achieved.
+
+positive
+: We do not support X11 directly on Ubuntu Core with Mir.
+The primary reason for this is security: the X11 protocol was not designed with security in mind, a malicious application connected to an X11 server can obtain much information from the other running X11 applications.
+
+Beware that not all toolkits have native support for Wayland. So, depending on the graphical toolkit your application uses, there may be some additional setup (for Xwayland) required.
+
+![graphical-snaps-with-mir](images/graphical-snaps-with-mir.png)
+
+### Toolkits with Native support for Wayland
+
+* GTK3/4 
+* Qt5
+* SDL2
+
+This is not an exhaustive list. There may be other toolkits that can work with Wayland but we know these work with Mir.
+ 
+Native support for Wayland the simplest case, as the application can talk to Mir directly.
+
+### Toolkits with No Wayland support
+
+* GTK2
+* Qt4
+* Electron apps
+* Java apps
+* Chromium
+
+This is a more complex case, as the toolkits require the legacy X11 protocol to function.
+
+To enable these applications we add an intermediary “Xwayland” which translates X11 calls to Wayland ones. Each snapped X11 application will have its own confined X11 server (Xwayland) which then talks Wayland - a far more secure protocol.
+
+positive
+: The current tutorial covers the development of all IoT graphical snaps. It is a prerequsite to [Graphical Snaps: Using Xwayland](tutorial/graphical-snaps-xwayland)  which explains the additional steps needed to run Xwayland in your application snap.
+
+## Preparation
+duration: 10:00
+
+### Requirements
+
+* An Ubuntu desktop (Xenial or Bionic)
```

Comment:
Is this really a requirement? Only for local testing I believe? Could we say that this was "tested on...", but should work cross-platform, and then later in the tutorial mention where things would only work on Ubuntu?

Reply:
Yes, it is the local testing that makes this a requirement. FWIW It shouldn't be hard to do that on Fedora too (if snapd is installed and slightly different install commands used).

However, for the initial cycle of learning it is wrong to offer options, the student doesn't have the knowledge to decide and is simply distracted. My judgment here is that documenting "the one true path" is best. (Once the initial phase is past, the student can apply the lessons in other contexts.)

Reply:
Ack, fine by me.